okhttp/src/main/java/okhttp3/Dispatcher.java

```diff
@@ -126,13 +140,16 @@ public synchronized void setIdleCallback(@Nullable Runnable idleCallback) {
     this.idleCallback = idleCallback;
   }
 
-  synchronized void enqueue(AsyncCall call) {
-    if (runningAsyncCalls.size() < maxRequests && runningCallsForHost(call) < maxRequestsPerHost) {
+  void enqueue(AsyncCall call) {
+    synchronized (this) {
+      if (runningAsyncCalls.size() >= maxRequests
```

Comment:
It could be simpler to just add to readyAsyncCalls and then call something like a new promoteAndExecute()

Avoids duplicating this logic but at some marginal cost on the happy path.  So just a thought.

Reply:
That worked nicely. Please consider taking another look!